earthfile2llb/wait_block.go

```diff
@@ -147,13 +149,47 @@ func (wb *waitBlock) saveImages(ctx context.Context) error {
 				}
 				platformImgNames[platformImgName] = true
 			}
+		} else {
+			if item.si.CheckDuplicate && item.si.DockerTag != "" {
+				if _, found := singPlatImgNames[item.si.DockerTag]; found {
+					return errors.Errorf(
+						"image %s is defined multiple times for the same default platform",
+						item.si.DockerTag)
+				}
+				singPlatImgNames[item.si.DockerTag] = true
+			}
 		}
 
-		_, err = gwCrafter.AddPushImageEntry(ref, refID, item.si.DockerTag, true, item.si.InsecurePush, item.si.Image, platformBytes)
+		refPrefix, err := gwCrafter.AddPushImageEntry(ref, refID, item.si.DockerTag, item.push, item.si.InsecurePush, item.si.Image, platformBytes)
 		if err != nil {
 			return err
 		}
 		refID++
+
+		if item.localExport {
+			if isMultiPlatform[item.si.DockerTag] {
+				// local docker instance does not support multi-platform images, so we must create a new entry and set it to the platformImgName
+				refPrefix, err := gwCrafter.AddPushImageEntry(ref, refID, platformImgName, false, false, item.si.Image, nil)
+				if err != nil {
+					return err
+				}
+				if item.c.opt.UseLocalRegistry {
+					localRegPullID := pullPingMap.Insert(sessionID, platformImgName)
+					gwCrafter.AddMeta(fmt.Sprintf("%s/export-image-local-registry", refPrefix), []byte(localRegPullID))
```

Comment:
this logic is duplicated between builder.go and here; I can spend some more time looking into moving it into the gatewaycrafter if you think that's worthwhile.

Reply:
I think we should eliminate the duplication soon - but it doesn't have to be in this very PR.

BTW, also ideally nothing adds meta entries or refs to the crafter directly - that part is abstracted away by the crafter. The crafter just offers a way to add images and artifacts (+ settings about how to export them) and deals with what metadata needs to be set accordingly within the implementation, without exposing it.

Reply:
That's the long-term vision indeed, eventually I want to remove the refPrefix it returns.